modexp/src/lib.rs

```diff
@@ -0,0 +1,335 @@
+extern crate ewasm_api;
+extern crate num_bigint;
+
+use num::{One, Zero};
+use num_bigint::{BigInt, Sign};
+use std::cmp;
+use std::ops::{Rem, Shl};
+
+fn f(x: usize) -> u64 {
+    let cost = match x {
+        0...64 => x * x,
+        65...1024 => (x * x / 4) + 96 * x - 3072,
+        _ => (x * x / 16) + 480 * x - 199680,
+    };
+    cost as u64
+}
+
+const G_QUAD_DIVISOR: u64 = 20;
+
+fn calculate_cost(lm: usize, lb: usize, exp: &BigInt) -> u64 {
+    let le = exp.to_bytes_be().1.len();
+    let lep: u64 = match le {
+        0...32 => {
+            if exp.is_zero() {
+                0u64
+            } else {
+                // Number of bits == ⌊log2()⌋ + 1
+                exp.bits() as u64 - 1
+            }
+        }
+        _ => {
+            let nbits = exp.bits();
+            let (loglow32bytes, low32bytes) = if nbits > 256 {
+                let x = exp.rem(&BigInt::one().shl(256));
+                (x.bits() - 1, x)
+            } else {
+                (nbits - 1, exp.clone())
+            };
+
+            if low32bytes > BigInt::zero() {
+                8 * (le as u64 - 32) + loglow32bytes as u64
+            } else {
+                8 * (le as u64 - 32)
+            }
+        }
+    };
+
+    (f(cmp::max(lm, lb)) * cmp::max(lep, 1 as u64)) / G_QUAD_DIVISOR
+}
+
+fn modexp(base: &BigInt, exp: &BigInt, modulus: &BigInt) -> Vec<u8> {
+    let x = if modulus.is_zero() {
+        BigInt::zero()
+    } else {
+        base.modpow(exp, modulus)
+    };
+    let (_, mut data) = x.to_bytes_be();
+
+    /* padded on 32 bytes */
+    let mut padded = vec![0u8; (32 - (data.len() % 32)) % 32];
+    padded.append(&mut data);
+    padded
+}
+
+#[cfg(not(test))]
+#[no_mangle]
+pub extern "C" fn main() {
+    use num::cast::ToPrimitive;
+    use std::cmp::min;
+
+    const BASE_LENGTH_OFFSET: usize = 0;
+    const BASE_LENGTH_LENGTH: usize = 32;
+    const EXP_LENGTH_OFFSET: usize = BASE_LENGTH_LENGTH;
+    const EXP_LENGTH_LENGTH: usize = 32;
+    const MOD_LENGTH_OFFSET: usize = EXP_LENGTH_OFFSET + EXP_LENGTH_LENGTH;
+    const MOD_LENGTH_LENGTH: usize = 32;
+    const BASE_OFFSET: usize = MOD_LENGTH_OFFSET + MOD_LENGTH_LENGTH;
```

Comment:
I think it would be better to define all the constants in one place, e.g. at the top of the file or inside the relevant funcs.
For readability.